delta-lake/delta-spark341db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuLowShuffleMergeCommand.scala

```diff
@@ -0,0 +1,1083 @@
+        matchedExprs(idx))
+      new Column(Alias(caseWhen, col.name)())
+    }
+
+    val modifiedDF = {
+
+      // Make this a udf to avoid catalyst to be too aggressive to even remove the join!
+      val noopRowDroppedCol = udf(new GpuDeltaNoopUDF()).apply(!col(ROW_DROPPED_COL))
+
+      val modifiedDF = joinedDF.select(modifiedCols: _*)
+        // This will not filter anything since they always return true, but we need to avoid
+        // catalyst from optimizing these udf
+        .filter(noopRowDroppedCol && col(INCR_METRICS_COL))
+        .drop(ROW_DROPPED_COL, INCR_METRICS_COL, TARGET_ROW_PRESENT_COL, SOURCE_ROW_PRESENT_COL)
+
+      repartitionIfNeeded(modifiedDF)
+    }
+
+    modifiedDF
+  }
+
+  private def getUnmodifiedDF(touchedFiles: Map[String, (Roaring64Bitmap, AddFile)]): DataFrame = {
+    getTouchedTargetDF(touchedFiles)
+      .filter(!col(METADATA_ROW_DEL_COL))
+      .drop(TARGET_ROW_PRESENT_COL, METADATA_ROW_DEL_COL)
+  }
+}
+
+
+object MergeExecutor {
+
+  /**
+   * Spark UI will track all normal accumulators along with Spark tasks to show them on Web UI.
+   * However, the accumulator used by `MergeIntoCommand` can store a very large value since it
+   * tracks all files that need to be rewritten. We should ask Spark UI to not remember it,
+   * otherwise, the UI data may consume lots of memory. Hence, we use the prefix `internal.metrics.`
+   * to make this accumulator become an internal accumulator, so that it will not be tracked by
+   * Spark UI.
+   */
+  val TOUCHED_FILES_ACCUM_NAME = "internal.metrics.MergeIntoDelta.touchedFiles"
+
+  val ROW_ID_COL = "_row_id_"
+  val FILE_PATH_COL: String = GpuDeltaParquetFileFormatUtils.FILE_PATH_COL
+  val SOURCE_ROW_PRESENT_COL: String = "_source_row_present_"
+  val SOURCE_ROW_PRESENT_FIELD: StructField = StructField(SOURCE_ROW_PRESENT_COL, BooleanType,
+    nullable = false)
+  val TARGET_ROW_PRESENT_COL: String = "_target_row_present_"
+  val TARGET_ROW_PRESENT_FIELD: StructField = StructField(TARGET_ROW_PRESENT_COL, BooleanType,
+    nullable = false)
+  val ROW_DROPPED_COL: String = GpuDeltaMergeConstants.ROW_DROPPED_COL
+  val ROW_DROPPED_FIELD: StructField = StructField(ROW_DROPPED_COL, BooleanType, nullable = false)
+  val INCR_METRICS_COL: String = "_incr_metrics_"
+  val INCR_METRICS_FIELD: StructField = StructField(INCR_METRICS_COL, BooleanType, nullable = false)
+  val INCR_ROW_COUNT_COL: String = "_incr_row_count_"
+
+  // Some Delta versions use Literal(null) which translates to a literal of NullType instead
+  // of the Literal(null, StringType) which is needed, so using a fixed version here
+  // rather than the version from Delta Lake.
+  val CDC_TYPE_NOT_CDC_LITERAL: Literal = Literal(null, StringType)
+
+  def toDeletionVector(bitmap: Roaring64Bitmap): DeletionVectorDescriptor = {
+    DeletionVectorDescriptor.inlineInLog(RoaringBitmapWrapper(bitmap).serializeToBytes(),
+      bitmap.getLongCardinality)
```

Comment:
> Currently we can't use RoaringBitmapArray because we rely on [this api](https://javadoc.io/static/org.roaringbitmap/RoaringBitmap/0.9.25/org/roaringbitmap/longlong/Roaring64Bitmap.html#getLongIteratorFrom(long)).

I don't see why we couldn't build an interator for RoaringBitmapArray that walks the bitmaps and produces the corresponding long values.  Seems relatively straightforward, but agree not necessary for this short-term prototype, especially if we move it to the GPU.

> We are not blindly assuming it.

There is a flag for whether the bitmap is inlined or not, but the code is unconditionally calling the method to get the inlined bitmap.  If the bitmap is not inlined, the query crashes.  That's what I meant by blindly assuming.

>  The way they distinguish it is to add a 4-byte magic number as the header of bytes. We can extend this when we acutally want to support deletion vector on read.

Sounds reasonable.